### PR TITLE
Don't allow withdraw milestone when balance is less than limit

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to #1885 and needs Giveth/feathers-giveth#434

### CODE REVIEW

File: src/components/MilestoneActions.jsx [hunk 1]
```diff
@@ -13,15 +13,72 @@ import ApproveRejectMilestoneCompletionButtons from 'components/ApproveRejectMil
...
+
+    if (
+      !currentBalanceUsdValue &&
+      currentUser.address &&
+      milestone.donationCounters &&
+      milestone.donationCounters.length
+    ) {
+      calculateMilestoneCurrentBalanceUsdValue();
+    }
+  });
```
Comment (aminlatifi) [inline src/components/MilestoneActions.jsx:67]: This should be dependent on milestone

File: src/components/MilestoneActions.jsx [hunk 2]
```diff
@@ -13,15 +13,72 @@ import ApproveRejectMilestoneCompletionButtons from 'components/ApproveRejectMil
...
     <Fragment>
       <AcceptRejectProposedMilestoneButtons milestone={milestone} />
 
       <ReproposeRejectedMilestoneButton milestone={milestone} />
 
-      {milestone.hasRecipient ? <RequestMarkMilestoneCompleteButton milestone={milestone} /> : null}
+      {milestone.hasRecipient ? (
+        <RequestMarkMilestoneCompleteButton
+          isAmountEnoughForWithdraw={isAmountEnoughForWithdraw}
+          minimumPayoutUsdValue={minimumPayoutUsdValue}
```
Comment (aminlatifi) [inline src/components/MilestoneActions.jsx:78]: It's not necessary to pass it, `RequestMarkMilestoneCompleteButton` can use context itself

File: src/components/MilestoneActions.jsx [hunk 3]
```diff
@@ -33,7 +90,11 @@ function MilestoneActions({ milestone }) {
 
       <ApproveRejectMilestoneCompletionButtons milestone={milestone} />
 
-      <WithdrawMilestoneFundsButton milestone={milestone} />
+      <WithdrawMilestoneFundsButton
+        milestone={milestone}
+        isAmountEnoughForWithdraw={isAmountEnoughForWithdraw}
+        minimumPayoutUsdValue={minimumPayoutUsdValue}
```
Comment (aminlatifi) [inline src/components/MilestoneActions.jsx:96]: Here too

File: src/components/RequestMarkMilestoneCompleteButton.jsx [hunk 4]
```diff
@@ -27,6 +31,16 @@ const RequestMarkMilestoneCompleteButton = ({ milestone }) => {
     actionWithLoggedIn(currentUser).then(() =>
       checkBalance(balance)
         .then(async () => {
+          if (!isAmountEnoughForWithdraw) {
+            ErrorPopup(
```
Comment (aminlatifi) [inline src/components/RequestMarkMilestoneCompleteButton.jsx:35]: Use new ant design message
https://ant.design/components/message/

File: src/components/WithdrawMilestoneFundsButton.jsx [hunk 5]
```diff
@@ -24,13 +28,22 @@ const WithdrawMilestoneFundsButton = ({ milestone }) => {
...
     const isRecipient = milestone.recipientAddress === userAddress;
-
     actionWithLoggedIn(currentUser).then(() =>
       Promise.all([
         checkBalance(balance),
         DonationService.getMilestoneDonationsCount(milestone._id),
       ])
         .then(([, donationsCount]) => {
+          if (!isAmountEnoughForWithdraw) {
+            ErrorPopup(
```
Comment (aminlatifi) [inline src/components/WithdrawMilestoneFundsButton.jsx:38]: As above use antd message

File: src/components/ViewMilestoneAlerts.jsx [hunk 6]
```diff
@@ -85,6 +98,8 @@ const ViewMilestoneAlerts = ({ milestone, campaign }) => {
 ViewMilestoneAlerts.propTypes = {
   milestone: PropTypes.instanceOf(Milestone).isRequired,
   campaign: PropTypes.instanceOf(Campaign).isRequired,
+  isAmountEnoughForWithdraw: PropTypes.bool.isRequired,
+  minimumPayoutUsdValue: PropTypes.number.isRequired,
```
Comment (aminlatifi) [inline src/components/ViewMilestoneAlerts.jsx:102]: Don't get it by property

File: src/components/WithdrawMilestoneFundsButton.jsx [hunk 7]
```diff
@@ -156,6 +169,8 @@ const WithdrawMilestoneFundsButton = ({ milestone }) => {
 
 WithdrawMilestoneFundsButton.propTypes = {
   milestone: PropTypes.instanceOf(Milestone).isRequired,
+  isAmountEnoughForWithdraw: PropTypes.bool.isRequired,
+  minimumPayoutUsdValue: PropTypes.number.isRequired,
```
Comment (aminlatifi) [inline src/components/WithdrawMilestoneFundsButton.jsx:173]: Don't get it by property

File: src/components/views/EditMilestone.jsx [hunk 8]
```diff
@@ -567,7 +567,7 @@ class EditMilestone extends Component {
       return Promise.reject();
     }
 
-    return authenticateUser(this.props.currentUser, true)
+    return authenticateIfPossible(this.props.currentUser, true)
```
Comment (aminlatifi) [inline src/components/views/EditMilestone.jsx:570]: This change is not necessary, `authenticateIfPossible` is identical and should be removed from middleware.jsx

**Comment by aminlatifi** [comment]
@RamRamez Please apply changed I noted in review

**Comment by RamRamez** [comment]
> @RamRamez Please apply changes I noted in review

Done

**Comment by aminlatifi** [comment]
@RamRamez Please test the functionality of this PR too

**Comment by RamRamez** [comment]
> @RamRamez Please test the functionality of this PR too

It's ok for me, but when I click "Disburse" once, the notification of "Oh No! A minimum donation balance of ..." appears behind the main menu and it's invisible to the user and he may not understand why the button is not working.

**Comment by aminlatifi** [comment]
> > @RamRamez Please test the functionality of this PR too
> 
> It's ok for me, but when I click "Disburse" once, the notification of "Oh No! A minimum donation balance of ..." appears behind the main menu and it's invisible to the user and he may not understand why the button is not working.

Thanks @RamRamez , the notification error is solved.

**Pull request merged**